crates/ra_hir_ty/src/expr.rs

```diff
@@ -312,3 +386,132 @@ pub fn record_pattern_missing_fields(
     }
     Some((variant_def, missed_fields, exhaustive))
 }
+
+#[cfg(test)]
+mod tests {
+    use insta::assert_snapshot;
+    use ra_db::fixture::WithFixture;
+
+    use crate::{diagnostics::MismatchedArgCount, test_db::TestDB};
+
+    fn check_diagnostic_message(ra_fixture: &str) -> String {
+        TestDB::with_single_file(ra_fixture).0.diagnostic::<MismatchedArgCount>().0
+    }
+
+    fn check_no_diagnostic(ra_fixture: &str) {
+        let (s, diagnostic_count) =
+            TestDB::with_single_file(ra_fixture).0.diagnostic::<MismatchedArgCount>();
+
+        assert_eq!(0, diagnostic_count, "expected no diagnostic, found one: {}", s);
+    }
+
+    #[test]
+    fn simple_free_fn_zero() {
+        assert_snapshot!(check_diagnostic_message(
+            r"
+            fn zero() {}
+
+            fn f() {
+                zero(1);
+            }
```

Comment:
I suggest using a more "compressed" style for tests: 

```rust
fn zero() {}
fn main() { zero(1); }
```

Seems like for small snippets extra newlines don't increase visibility, but they increase the overall percieved verbosity of tests